core/trino-spi/src/main/java/io/trino/spi/connector/ConnectorPageSinkProvider.java

```diff
@@ -21,15 +22,35 @@
 {
     ConnectorPageSink createPageSink(ConnectorTransactionHandle transactionHandle, ConnectorSession session, ConnectorOutputTableHandle outputTableHandle);
 
+    default ConnectorPageSink createPageSink(ConnectorTransactionHandle transactionHandle, ConnectorSession session, ConnectorOutputTableHandle outputTableHandle, TaskId taskId)
```

Comment:
passing `TaskId` to connectors feels like exposing too much of internal details of the engine to the connectors. 
I do not feel strongly agains but I would like to understand why not allow connector to use UUID internally (I did not read that part of PR but I guess you are using taskID to tag the rows in the temporary table used for inserts, to later filter those out). Is it just about fact that taskId is smaller? Do we care really that much?

cc: @arhimondr as I guess you were involved in the design dicussion

Reply:
Also if we decide to keep the SPI change it should be a separate commit.

Reply:
@arhimondr and i debated for a while about using UUIDs - that was indeed the first thought we went with. 

We didn't like UUIDs because we'd have to use a `char(36)` field to store them. Not only was storage space a concern at very large scale, but the join speed of such a field is not as good as with an `int` or `bigint`.

We considered using only one `long` from a UUID, but we were concerned about the possibility of a collision - however slight that might be. I pointed out that if two tasks chose the same `long`, we could turn the `Collection<Slice>` passed to the coordinator into a `Set<Long>`, and if its size was smaller than the `Collection<Slice>`, that would detect a collision, and we'd fail the query. 

Obviously, failing the query at the `finish` stage is not great, but more importantly, Andrii pointed out that a successful task's id could collide with a failed task's id, and if the failed task had managed to insert _some_ rows, the query results would be incorrect (or possibly cause a unique constraint violation), as we'd grab the rows from both the successful and failed task.

We decided we wanted to use the `TaskId`, as it guarantees a unique value (within the same query) for every attempt of every task. After a little bit more brainstorming, we were satisfied with the `long` constructed as i described above.

The SPI changes are indeed separate commits - would they need to be separate PRs?

Reply:
> @arhimondr and i debated for a while about using UUIDs - that was indeed the first thought we went with.
> 
> We didn't like UUIDs because we'd have to use a `char(36)` field to store them. Not only was storage space a concern at very large scale, but the join speed of such a field is not as good as with an `int` or `bigint`.
> 
> We considered using only one `long` from a UUID, but we were concerned about the possibility of a collision - however slight that might be. I pointed out that if two tasks chose the same `long`, we could turn the `Collection<Slice>` passed to the coordinator into a `Set<Long>`, and if its size was smaller than the `Collection<Slice>`, that would detect a collision, and we'd fail the query.
> 
> Obviously, failing the query at the `finish` stage is not great, but more importantly, Andrii pointed out that a successful task's id could collide with a failed task's id, and if the failed task had managed to insert _some_ rows, the query results would be incorrect (or possibly cause a unique constraint violation), as we'd grab the rows from both the successful and failed task.

IDK - I kinda would not extend SPI if not needed. I do not expect that we will be putting an enormous number of rows over single-threaded JDBC - so performance degradation should not be a big deal.
But if we really care about performance and extra storage another option would be to go with UUID and model it as two columns in temporary table. WDYT?

>The SPI changes are indeed separate commits - would they need to be separate PRs?

Not - separate commits is enough. Though changes to `core/trino-spi/src/main/java/io/trino/spi/connector/ConnectorPageSinkProvider.java` are not separated out.

 decided we wanted to use the `TaskId`, as it guarantees a unique value (within the same query) for every attempt of every task. After a little bit more brainstorming, we were satisfied with the `long` constructed as i described above.

Reply:
> putting an enormous number of rows over single-threaded JDBC - so performance degradation should not be a big deal

It's more about the performance of the final `insert into ... select from where <semi join>` - that semi join is more performant with a `bigint` than a `char(36)`.

i was also surprised that TaskId wasn't actually passed to the page sinks. It sounds like you spoke with andrii offline though and we're moving forward with this approach. please correct me if i'm wrong